File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* helper(ListNode* head,int k,int doRev){
        if(head==NULL) return NULL;
        if(doRev==0) return head;
        ListNode* hold=head;
        ListNode* cur=head;
        ListNode* prev=NULL;
        ListNode* nxt=NULL;
        int times=k;
        while(times-- && cur){
            nxt=cur->next;
            cur->next=prev;
            prev=cur;
            cur=nxt;
            doRev--;
        }
        hold->next=helper(cur,k,doRev);
        return prev;

    }
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* temp=head;
        int len=0;
        while(temp){
            temp=temp->next;
            len++;
        }
        int left=len%k;
        int doRev=len-left;
        cout<<doRev;
        return helper(head,k,doRev);
    }
};

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode helper(ListNode head,int k){
        ListNode temp=head;
        int count=0;
        while(temp!=null){
            count++;
            temp=temp.next;
        }
        if(count<k) return head;
        ListNode cur=head;
        ListNode next=null;
        ListNode prev=null;
        int times=k;
        while(cur!=null && k>=1){
            next=cur.next;
            cur.next=prev;
            prev=cur;
            cur=next;
            k--;
        }
        head.next=helper(next,times);
        return prev;
    }
    public ListNode ReverseKGroup(ListNode head, int k) {
        return helper(head,k);
    }
}